core/vdbe/mod.rs

```diff
@@ -3084,41 +3129,56 @@ impl Program {
         }
     }
 
-    fn halt(&self, pager: Rc<Pager>, program_state: &mut ProgramState) -> Result<StepResult> {
-        let connection = self
-            .connection
-            .upgrade()
-            .expect("only weak ref to connection?");
-        let auto_commit = *connection.auto_commit.borrow();
-        tracing::trace!("Halt auto_commit {}", auto_commit);
-        assert!(
-            program_state.halt_state.is_none()
-                || (matches!(program_state.halt_state.unwrap(), HaltState::Checkpointing))
-        );
-        if program_state.halt_state.is_some() {
-            self.step_end_write_txn(&pager, &mut program_state.halt_state, connection.deref())
+    fn halt(
+        &self,
+        pager: Rc<Pager>,
+        program_state: &mut ProgramState,
+        mv_store: Option<Rc<MvStore>>,
+    ) -> Result<StepResult> {
+        if let Some(mv_store) = mv_store {
+            let conn = self.connection.upgrade().unwrap();
+            let mut mv_transactions = conn.mv_transactions.borrow_mut();
+            for tx_id in mv_transactions.iter() {
+                mv_store.commit_tx(*tx_id).unwrap();
+            }
+            mv_transactions.clear();
```

Comment:
Halt may not commit, I imagine this case is missed here. I refer to `if !auto_commit` case